File: src-tauri/src/log_viewer.rs
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::sync::{Arc, LazyLock, Mutex};
use std::time::Duration;

use chrono::{DateTime, Local, NaiveDate, NaiveDateTime, NaiveTime, TimeZone, Utc};
use regex::Regex;
use serde::{Deserialize, Serialize};
use tauri::ipc::Channel;
use tauri::{AppHandle, Manager, Runtime};
use tokio::fs;
use tokio::io::{AsyncReadExt, AsyncSeekExt, SeekFrom};
use tokio::time::sleep;

const DEFAULT_MAX_INITIAL_LINES: usize = 500;
const DEFAULT_MAX_INITIAL_BYTES: usize = 1024 * 1024;
const DEFAULT_POLL_INTERVAL_MS: u64 = 500;
const MESSAGE_PREVIEW_LEN: usize = 200;
const MAX_LINE_LENGTH: usize = 50_000;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
#[serde(rename_all = "camelCase")]
pub enum LogSource {
    App,
    AgentServer,
    AiConversations,
}

impl LogSource {
    fn file_name(&self) -> &'static str {
        match self {
            LogSource::App => "kenvo.log",
            LogSource::AgentServer => "agent-server.log",
            LogSource::AiConversations => "ai-conversations.jsonl",
        }
    }

    fn dir_kind(&self) -> &'static str {
        match self {
            LogSource::App | LogSource::AgentServer => "log",
            LogSource::AiConversations => "data",
        }
    }

    fn path<R: Runtime>(&self, app: &AppHandle<R>) -> Result<PathBuf, String> {
        let dir = match self.dir_kind() {
            "log" => app.path().app_log_dir(),
            "data" => app.path().app_data_dir(),
            _ => unreachable!(),
        }
        .map_err(|e| e.to_string())?;
        Ok(dir.join(self.file_name()))
    }
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LogLine {
    pub source: LogSource,
    pub timestamp: String,
    pub timestamp_ms: i64,
    pub level: String,
    pub target: Option<String>,
    pub message: String,
    pub raw: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase", tag = "event", content = "data")]
pub enum LogStreamEvent {
    Initial { source: LogSource, lines: Vec<LogLine> },
    NewLines { source: LogSource, lines: Vec<LogLine> },
    Reset { source: LogSource, lines: Vec<LogLine> },
    Error { source: LogSource, message: String },
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamOptions {
    pub max_initial_lines: Option<usize>,
    pub tail: Option<bool>,
}

#[derive(Default)]
pub struct LogStreamState {
    cancel_flags: Mutex<HashMap<String, Arc<AtomicBool>>>,
    next_id: AtomicU64,
}

impl LogStreamState {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn insert(&self) -> (String, Arc<AtomicBool>) {
        let id = format!("log-stream-{}", self.next_id.fetch_add(1, Ordering::SeqCst));
        let flag = Arc::new(AtomicBool::new(false));
        self.cancel_flags.lock().unwrap().insert(id.clone(), flag.clone());
        (id, flag)
    }

    pub fn stop(&self, id: &str) -> bool {
        if let Some(flag) = self.cancel_flags.lock().unwrap().remove(id) {
            flag.store(true, Ordering::Relaxed);
            true
        } else {
            false
        }
    }

    pub fn remove(&self, id: &str) {
        self.cancel_flags.lock().unwrap().remove(id);
    }
}

#[tauri::command]
pub async fn stream_log<R: Runtime>(
    app: AppHandle<R>,
    state: tauri::State<'_, LogStreamState>,
    channel: Channel<LogStreamEvent>,
    sources: Vec<LogSource>,
    options: StreamOptions,
) -> Result<String, String> {
    let (stream_id, cancel) = state.insert();
    let stream_id_for_task = stream_id.clone();
    let app_for_task = app.clone();

    tauri::async_runtime::spawn(async move {
        let result = run_stream(app_for_task.clone(), &channel, &sources, options, &cancel).await;
        if let Err(e) = result {
            let _ = channel.send(LogStreamEvent::Error {
                source: LogSource::App,
                message: e,
            });
        }
        app_for_task
            .state::<LogStreamState>()
            .remove(&stream_id_for_task);
    });

    Ok(stream_id)
}

#[tauri::command]
pub fn stop_log_stream(state: tauri::State<'_, LogStreamState>, stream_id: String) -> Result<(), String> {
    state.stop(&stream_id);
    Ok(())
}

#[tauri::command]
pub async fn export_log<R: Runtime>(
    app: AppHandle<R>,
    source: LogSource,
    dest_path: String,
) -> Result<(), String> {
    let src = source.path(&app)?;
    fs::copy(&src, &dest_path)
        .await
        .map_err(|e| format!("failed to export log: {e}"))?;
    Ok(())
}

async fn run_stream<R: Runtime>(
    app: AppHandle<R>,
    channel: &Channel<LogStreamEvent>,
    sources: &[LogSource],
    options: StreamOptions,
    cancel: &Arc<AtomicBool>,
) -> Result<(), String> {
    let max_initial_lines = options.max_initial_lines.unwrap_or(DEFAULT_MAX_INITIAL_LINES);
    let tail = options.tail.unwrap_or(true);
    let interval = Duration::from_millis(DEFAULT_POLL_INTERVAL_MS);

    let mut source_states: Vec<(LogSource, PathBuf, u64, String)> = Vec::with_capacity(sources.len());

    for source in sources {
        let path = source.path(&app)?;
        match read_tail_lines(&path, DEFAULT_MAX_INITIAL_BYTES, max_initial_lines).await {
            Ok(lines) => {
                let parsed: Vec<LogLine> = lines
                    .iter()
                    .filter_map(|line| parse_line(*source, line))
                    .collect();
                if channel
                    .send(LogStreamEvent::Initial {
                        source: *source,
                        lines: parsed,
                    })
                    .is_err()
                {
                    return Ok(());
                }
                let size = fs::metadata(&path).await.map(|m| m.len()).unwrap_or(0);
                source_states.push((*source, path, size, String::new()));
            }
            Err(e) => {
                let _ = channel.send(LogStreamEvent::Error {
                    source: *source,
                    message: e,
                });
                source_states.push((*source, path, 0, String::new()));
            }
        }
    }

    if !tail {
        return Ok(());
    }

    loop {
        sleep(interval).await;
        if cancel.load(Ordering::Relaxed) {
            break;
        }

        for (source, path, pos, pending) in &mut source_states {
            match read_new_data(path, *pos, pending).await {
                Ok((lines, new_pos, rotated)) => {
                    *pos = new_pos;
                    if lines.is_empty() {
                        continue;
                    }
                    let parsed: Vec<LogLine> = lines
                        .iter()
                        .filter_map(|line| parse_line(*source, line))
                        .collect();
                    let event = if rotated {
                        LogStreamEvent::Reset {
                            source: *source,
                            lines: parsed,
                        }
                    } else {
                        LogStreamEvent::NewLines {
                            source: *source,
                            lines: parsed,
                        }
                    };
                    if channel.send(event).is_err() {
                        return Ok(());
                    }
                }
                Err(e) => {
                    let _ = channel.send(LogStreamEvent::Error {
                        source: *source,
                        message: e,
                    });
                }
            }
        }
    }

    Ok(())
}

async fn read_tail_lines(
    path: &Path,
    max_bytes: usize,
    max_lines: usize,
) -> Result<Vec<String>, String> {
    let metadata = fs::metadata(path)
        .await
        .map_err(|e| format!("failed to read log metadata: {e}"))?;
    let size = metadata.len();
    if size == 0 {
        return Ok(Vec::new());
    }

    let start = if size <= max_bytes as u64 {
        0
    } else {
        size - max_bytes as u64
    };

    let mut file = fs::File::open(path)
        .await
        .map_err(|e| format!("failed to open log file: {e}"))?;
    file.seek(SeekFrom::Start(start))
        .await
        .map_err(|e| format!("failed to seek log file: {e}"))?;

    let mut buf = vec![0u8; (size - start) as usize];
    file.read_exact(&mut buf)
        .await
        .map_err(|e| format!("failed to read log file: {e}"))?;

    let text = String::from_utf8_lossy(&buf);
    let mut lines: Vec<String> = text.split('\n').map(|s| s.to_string()).collect();
    if start > 0 && !lines.is_empty() {
        lines.remove(0);
    }
    if lines.len() > max_lines {
        let split_at = lines.len() - max_lines;
        lines = lines.split_off(split_at);
    }
    Ok(lines)
}

async fn read_new_data(
    path: &Path,
    last_pos: u64,
    pending: &mut String,
) -> Result<(Vec<String>, u64, bool), String> {
    let metadata = fs::metadata(path)
        .await
        .map_err(|e| format!("failed to read log metadata: {e}"))?;
    let size = metadata.len();

    if size < last_pos {
        pending.clear();
        let lines = read_tail_lines(path, DEFAULT_MAX_INITIAL_BYTES, DEFAULT_MAX_INITIAL_LINES).await?;
        return Ok((lines, size, true));
    }

    if size == last_pos {
        return Ok((Vec::new(), last_pos, false));
    }

    let mut file = fs::File::open(path)
        .await
        .map_err(|e| format!("failed to open log file: {e}"))?;
    file.seek(SeekFrom::Start(last_pos))
        .await
        .map_err(|e| format!("failed to seek log file: {e}"))?;

    let mut buf = Vec::with_capacity((size - last_pos) as usize);
    file.read_to_end(&mut buf)
        .await
        .map_err(|e| format!("failed to read log file: {e}"))?;

    let text = String::from_utf8_lossy(&buf);
    let combined = std::mem::take(pending) + &text;
    let mut lines: Vec<String> = combined.split('\n').map(|s| s.to_string()).collect();

    if combined.ends_with('\n') {
        *pending = String::new();
    } else if let Some(last) = lines.pop() {
        *pending = last;
    } else {
        *pending = String::new();
    }

    Ok((lines, size, false))
}

static APP_LOG_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^\[(\d{4}-\d{2}-\d{2})\]\[(\d{2}:\d{2}:\d{2}(?:\.\d+)?)\]\[(\w+)\]\[(.*?)\]\s*(.*)$")
        .unwrap()
});

static AGENT_LOG_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})?)\s+\[(\w+)\]\s+(.*)$")
        .unwrap()
});

fn parse_line(source: LogSource, raw: &str) -> Option<LogLine> {
    let trimmed = raw.trim_end_matches('\r');
    if trimmed.trim().is_empty() {
        return None;
    }
    let raw = if raw.len() > MAX_LINE_LENGTH {
        format!("{}...(truncated)", &raw[..MAX_LINE_LENGTH])
    } else {
        raw.to_string()
    };

    match source {
        LogSource::App => parse_app_line(&raw),
        LogSource::AgentServer => parse_agent_line(&raw),
        LogSource::AiConversations => parse_ai_line(&raw),
    }
}

fn parse_app_line(raw: &str) -> Option<LogLine> {
    let caps = APP_LOG_RE.captures(raw)?;
    let date_str = caps.get(1)?.as_str();
    let time_str = caps.get(2)?.as_str();
    let level = caps.get(3)?.as_str().to_uppercase();
    let target = caps.get(4)?.as_str();
    let message = caps.get(5)?.as_str().to_string();

    let date = NaiveDate::parse_from_str(date_str, "%Y-%m-%d").ok()?;
    let time = NaiveTime::parse_from_str(time_str, "%H:%M:%S%.f").ok()?;
    let naive = NaiveDateTime::new(date, time);
    let dt = Local
        .from_local_datetime(&naive)
        .earliest()
        .unwrap_or_else(|| Local.from_utc_datetime(&naive));

    Some(LogLine {
        source: LogSource::App,
        timestamp: dt.to_rfc3339(),
        timestamp_ms: dt.timestamp_millis(),
        level,
        target: if target.is_empty() { None } else { Some(target.to_string()) },
        message: truncate_message(message),
        raw: raw.to_string(),
    })
}

fn parse_agent_line(raw: &str) -> Option<LogLine> {
    let caps = AGENT_LOG_RE.captures(raw)?;
    let ts_str = caps.get(1)?.as_str();
    let level = caps.get(2)?.as_str().to_uppercase();
    let rest = caps.get(3)?.as_str();

    let dt = DateTime::parse_from_rfc3339(ts_str)
        .or_else(|_| DateTime::parse_from_str(ts_str, "%+"))
        .ok()?;
    let dt_utc = dt.with_timezone(&Utc);

    let (message, target) = split_agent_rest(rest);

    Some(LogLine {
        source: LogSource::AgentServer,
        timestamp: dt_utc.to_rfc3339(),
        timestamp_ms: dt_utc.timestamp_millis(),
        level,
        target,
        message: truncate_message(message),
        raw: raw.to_string(),
    })
}

fn split_agent_rest(rest: &str) -> (String, Option<String>) {
    let trimmed = rest.trim_end();
    if let Some(brace_pos) = trimmed.rfind('{') {
        let (msg_part, json_part) = trimmed.split_at(brace_pos);
        if let Ok(json) = serde_json::from_str::<serde_json::Value>(json_part) {
            let target = json
                .get("agentId")
                .or_else(|| json.get("task"))
                .and_then(|v| v.as_str())
                .map(|s| s.to_string());
            return (msg_part.trim().to_string(), target);
        }
    }
    (trimmed.to_string(), None)
}

fn parse_ai_line(raw: &str) -> Option<LogLine> {
    let json = serde_json::from_str::<serde_json::Value>(raw).ok()?;
    let ts_str = json.get("ts")?.as_str()?;
    let dt = DateTime::parse_from_rfc3339(ts_str)
        .or_else(|_| DateTime::parse_from_str(ts_str, "%+"))
        .ok()?;
    let dt_utc = dt.with_timezone(&Utc);

    let ty = json.get("type").and_then(|v| v.as_str()).unwrap_or("unknown");
    let level = match ty {
        "error" => "ERROR",
        "warning" | "warn" => "WARN",
        "conversation" => "DEBUG",
        _ => "INFO",
    }
    .to_string();

    let operation_id = json
        .get("operationId")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    let provider = json.get("provider").and_then(|v| v.as_str()).unwrap_or("");
    let model = json.get("modelId").and_then(|v| v.as_str()).unwrap_or("");
    let call_id = json.get("callId").and_then(|v| v.as_str()).unwrap_or("");

    let message = if operation_id.is_empty() {
        format!("{} {}", ty, call_id)
    } else {
        let mut parts = vec![ty.to_string(), operation_id.to_string()];
        if !provider.is_empty() {
            parts.push(provider.to_string());
        }
        if !model.is_empty() {
            parts.push(model.to_string());
        }
        parts.join(" | ")
    };

    Some(LogLine {
        source: LogSource::AiConversations,
        timestamp: dt_utc.to_rfc3339(),
        timestamp_ms: dt_utc.timestamp_millis(),
        level,
        target: Some(ty.to_string()),
        message: truncate_message(message),
        raw: raw.to_string(),
    })
}

fn truncate_message(message: String) -> String {
    if message.len() > MESSAGE_PREVIEW_LEN {
        format!("{}...", &message[..MESSAGE_PREVIEW_LEN])
    } else {
        message
    }
}

File: apps/desktop/electron/log-viewer.ts
import fs from 'fs/promises'
import path from 'path'

import { WebContents } from 'electron'

import { dbExecute, dbSelect } from './db'
import { IPC_CHANNELS } from './ipc-channels'
import { getAppPaths } from './paths'

export type LogSource = 'app' | 'agentServer' | 'aiConversations'

export interface LogLine {
  source: LogSource
  timestamp: string
  timestampMs: number
  level: string
  target?: string
  message: string
  raw: string
}

export type LogStreamEvent =
  | { event: 'initial'; data: { source: LogSource; lines: LogLine[] } }
  | { event: 'newLines'; data: { source: LogSource; lines: LogLine[] } }
  | { event: 'reset'; data: { source: LogSource; lines: LogLine[] } }
  | { event: 'error'; data: { source: LogSource; message: string } }

export interface StreamOptions {
  maxInitialLines?: number
  tail?: boolean
}

export interface LogQueryOptions {
  afterId?: number
  limit?: number
  level?: string
  keyword?: string
  fromTimestampMs?: number
  toTimestampMs?: number
}

export interface StoredLogRecord {
  id: number
  source: LogSource
  timestampMs: number
  level: string
  target: string | null
  message: string
  raw: string
}

const DEFAULT_MAX_INITIAL_LINES = 500
const DEFAULT_MAX_INITIAL_BYTES = 1024 * 1024
const DEFAULT_POLL_INTERVAL_MS = 500
const DEFAULT_MAX_QUERY_LINES = 1000
const MESSAGE_PREVIEW_LEN = 200
const MAX_LINE_LENGTH = 50_000

const APP_LOG_RE =
  /^\[(\d{4}-\d{2}-\d{2})\]\[(\d{2}:\d{2}:\d{2}(?:\.\d+)?)\]\[(\w+)\]\[(.*?)\]\s*(.*)$/
const AGENT_LOG_RE =
  /^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})?)\s+\[(\w+)\]\s+(.*)$/

interface ActiveStream {
  cancel: boolean
  sender: WebContents
  interval?: NodeJS.Timeout
  pollInFlight?: boolean
  fallbackStarted?: boolean
}

const activeStreams = new Map<string, ActiveStream>()
let nextStreamId = 1

function sourceFileName(source: LogSource): string {
  switch (source) {
    case 'app':
      return 'kenvo.log'
    case 'agentServer':
      return 'agent-server.log'
    case 'aiConversations':
      return 'ai-conversations.jsonl'
  }
}

function sourceDir(source: LogSource): string {
  const { logDir, dataDir } = getAppPaths()
  return source === 'aiConversations' ? dataDir : logDir
}

function sourcePath(source: LogSource): string {
  return path.join(sourceDir(source), sourceFileName(source))
}

function send(sender: WebContents, event: LogStreamEvent): void {
  sender.send(IPC_CHANNELS.LOG_EVENT, event)
}

const VALID_SOURCES = new Set<LogSource>(['app', 'agentServer', 'aiConversations'])

function normalizeSources(sources: LogSource[]): LogSource[] {
  const unique = new Set<LogSource>()
  for (const source of sources) {
    if (VALID_SOURCES.has(source)) unique.add(source)
  }
  return [...unique]
}

function toLogLine(record: StoredLogRecord): LogLine {
  return {
    source: record.source,
    timestamp: new Date(record.timestampMs).toISOString(),
    timestampMs: record.timestampMs,
    level: record.level.toUpperCase(),
    target: record.target ?? undefined,
    message: truncateMessage(record.message),
    raw: record.raw,
  }
}

function queryRecordsForSource(
  source: LogSource,
  limit?: number,
  options: LogQueryOptions = {},
): StoredLogRecord[] {
  const conditions = ['source = ?']
  const params: unknown[] = [source]
  if (options.level) {
    conditions.push('level = ?')
    params.push(options.level.toUpperCase())
  }
  if (options.fromTimestampMs !== undefined) {
    conditions.push('timestamp_ms >= ?')
    params.push(options.fromTimestampMs)
  }
  if (options.toTimestampMs !== undefined) {
    conditions.push('timestamp_ms <= ?')
    params.push(options.toTimestampMs)
  }
  if (options.keyword?.trim()) {
    const keyword = `%${options.keyword.trim().toLowerCase()}%`
    conditions.push(
      "(lower(message) LIKE ? OR lower(COALESCE(target, '')) LIKE ? OR lower(raw) LIKE ?)",
    )
    params.push(keyword, keyword, keyword)
  }
  const limitSql = limit === undefined ? '' : ' LIMIT ?'
  if (limit !== undefined) params.push(limit)
  const rows = dbSelect<StoredLogRecord>(
    `SELECT id, source, timestamp_ms AS timestampMs, level, target, message, raw
       FROM logs
      WHERE ${conditions.join(' AND ')}
      ORDER BY id DESC${limitSql}`,
    params,
  )
  return rows.reverse()
}

function queryRecordsAfterId(
  sources: LogSource[],
  afterId: number,
  limit: number,
  options: LogQueryOptions = {},
): StoredLogRecord[] {
  const placeholders = sources.map(() => '?').join(', ')
  const conditions = [`source IN (${placeholders})`, 'id > ?']
  const params: unknown[] = [...sources, afterId]
  if (options.level) {
    conditions.push('level = ?')
    params.push(options.level.toUpperCase())
  }
  if (options.fromTimestampMs !== undefined) {
    conditions.push('timestamp_ms >= ?')
    params.push(options.fromTimestampMs)
  }
  if (options.toTimestampMs !== undefined) {
    conditions.push('timestamp_ms <= ?')
    params.push(options.toTimestampMs)
  }
  if (options.keyword?.trim()) {
    const keyword = `%${options.keyword.trim().toLowerCase()}%`
    conditions.push(
      "(lower(message) LIKE ? OR lower(COALESCE(target, '')) LIKE ? OR lower(raw) LIKE ?)",
    )
    params.push(keyword, keyword, keyword)
  }
  params.push(limit)
  return dbSelect<StoredLogRecord>(
    `SELECT id, source, timestamp_ms AS timestampMs, level, target, message, raw
       FROM logs
      WHERE ${conditions.join(' AND ')}
      ORDER BY id ASC
      LIMIT ?`,
    params,
  )
}

function queryInitialRecords(sources: LogSource[], maxInitialLines: number): StoredLogRecord[] {
  return sources.flatMap((source) => queryRecordsForSource(source, maxInitialLines))
}

export function queryLogRecords(sources: LogSource[], options: LogQueryOptions = {}): LogLine[] {
  const normalizedSources = normalizeSources(sources)
  if (normalizedSources.length === 0) return []

  const records =
    options.afterId === undefined
      ? normalizedSources.flatMap((source) =>
          queryRecordsForSource(source, options.limit ?? DEFAULT_MAX_INITIAL_LINES, options),
        )
      : queryRecordsAfterId(
          normalizedSources,
          options.afterId,
          options.limit ?? DEFAULT_MAX_QUERY_LINES,
          options,
        )
  return records
    .map(toLogLine)
    .sort((a, b) => a.timestampMs - b.timestampMs || a.source.localeCompare(b.source))
}

export function clearLogRecords(sources: LogSource[]): number {
  const normalizedSources = normalizeSources(sources)
  if (normalizedSources.length === 0) return 0
  const placeholders = normalizedSources.map(() => '?').join(', ')
  return dbExecute(`DELETE FROM logs WHERE source IN (${placeholders})`, normalizedSources).changes
}

function truncateMessage(message: string): string {
  if (message.length > MESSAGE_PREVIEW_LEN) {
    return `${message.slice(0, MESSAGE_PREVIEW_LEN)}...`
  }
  return message
}

function parseAppLine(raw: string): LogLine | null {
  const match = APP_LOG_RE.exec(raw)
  if (!match) return null
  const [, dateStr, timeStr, level, target, message] = match
  const ts = new Date(`${dateStr}T${timeStr}`)
  if (Number.isNaN(ts.getTime())) return null
  return {
    source: 'app',
    timestamp: ts.toISOString(),
    timestampMs: ts.getTime(),
    level: level.toUpperCase(),
    target: target || undefined,
    message: truncateMessage(message),
    raw,
  }
}

function splitAgentRest(rest: string): { message: string; target?: string } {
  const trimmed = rest.trimEnd()
  const bracePos = trimmed.lastIndexOf('{')
  if (bracePos >= 0) {
    const msgPart = trimmed.slice(0, bracePos)
    const jsonPart = trimmed.slice(bracePos)
    try {
      const json = JSON.parse(jsonPart)
      const target = (json.agentId || json.task || undefined) as string | undefined
      return { message: msgPart.trim(), target }
    } catch {
      // ignore
    }
  }
  return { message: trimmed }
}

function parseAgentLine(raw: string): LogLine | null {
  const match = AGENT_LOG_RE.exec(raw)
  if (!match) return null
  const [, tsStr, level, rest] = match
  const ts = new Date(tsStr)
  if (Number.isNaN(ts.getTime())) return null
  const { message, target } = splitAgentRest(rest)
  return {
    source: 'agentServer',
    timestamp: ts.toISOString(),
    timestampMs: ts.getTime(),
    level: level.toUpperCase(),
    target,
    message: truncateMessage(message),
    raw,
  }
}

function parseAiLine(raw: string): LogLine | null {
  try {
    const json = JSON.parse(raw)
    const tsStr = json.ts as string | undefined
    if (!tsStr) return null
    const ts = new Date(tsStr)
    if (Number.isNaN(ts.getTime())) return null

    const ty = (json.type as string) ?? 'unknown'
    const level =
      ty === 'error'
        ? 'ERROR'
        : ty === 'warning' || ty === 'warn'
          ? 'WARN'
          : ty === 'conversation'
            ? 'DEBUG'
            : 'INFO'

    const operationId = (json.operationId as string) ?? ''
    const provider = (json.provider as string) ?? ''
    const model = (json.modelId as string) ?? ''
    const callId = (json.callId as string) ?? ''

    const message = operationId
      ? [ty, operationId, provider, model].filter(Boolean).join(' | ')
      : `${ty} ${callId}`.trim()

    return {
      source: 'aiConversations',
      timestamp: ts.toISOString(),
      timestampMs: ts.getTime(),
      level,
      target: ty,
      message: truncateMessage(message),
      raw,
    }
  } catch {
    return null
  }
}

function parseLine(source: LogSource, raw: string): LogLine | null {
  const trimmed = raw.trimEnd().trim()
  if (!trimmed) return null
  const safeRaw =
    raw.length > MAX_LINE_LENGTH ? `${raw.slice(0, MAX_LINE_LENGTH)}...(truncated)` : raw
  switch (source) {
    case 'app':
      return parseAppLine(safeRaw)
    case 'agentServer':
      return parseAgentLine(safeRaw)
    case 'aiConversations':
      return parseAiLine(safeRaw)
  }
}

async function readTailLines(
  filePath: string,
  maxBytes: number,
  maxLines: number,
): Promise<string[]> {
  let statInfo: Awaited<ReturnType<typeof fs.stat>>
  try {
    statInfo = await fs.stat(filePath)
  } catch {
    return []
  }
  const size = statInfo.size
  if (size === 0) return []
  const start = size <= maxBytes ? 0 : size - maxBytes
  const handle = await fs.open(filePath, 'r')
  try {
    const buffer = Buffer.alloc(Number(size - start))
    await handle.read(buffer, 0, buffer.length, Number(start))
    let text = buffer.toString('utf-8')
    if (start > 0 && text.includes('\n')) {
      text = text.slice(text.indexOf('\n') + 1)
    }
    const lines = text.split('\n')
    if (lines.length > maxLines) {
      return lines.slice(lines.length - maxLines)
    }
    return lines
  } finally {
    await handle.close()
  }
}

async function readNewData(
  filePath: string,
  lastPos: number,
  pendingRef: { value: string },
): Promise<{ lines: string[]; newPos: number; rotated: boolean }> {
  let statInfo: Awaited<ReturnType<typeof fs.stat>>
  try {
    statInfo = await fs.stat(filePath)
  } catch (e) {
    throw new Error(`failed to read log metadata: ${(e as Error).message}`)
  }
  const size = statInfo.size

  if (size < lastPos) {
    pendingRef.value = ''
    const lines = await readTailLines(
      filePath,
      DEFAULT_MAX_INITIAL_BYTES,
      DEFAULT_MAX_INITIAL_LINES,
    )
    return { lines, newPos: size, rotated: true }
  }

  if (size === lastPos) {
    return { lines: [], newPos: lastPos, rotated: false }
  }

  const handle = await fs.open(filePath, 'r')
  try {
    const buffer = Buffer.alloc(Number(size - lastPos))
    await handle.read(buffer, 0, buffer.length, Number(lastPos))
    const text = buffer.toString('utf-8')
    const combined = pendingRef.value + text
    pendingRef.value = ''
    const lines = combined.split('\n')
    if (!combined.endsWith('\n')) {
      pendingRef.value = lines.pop() ?? ''
    }
    return { lines, newPos: size, rotated: false }
  } finally {
    await handle.close()
  }
}

interface SourceState {
  source: LogSource
  path: string
  position: number
  pending: { value: string }
}

function runDatabaseStream(
  streamId: string,
  sender: WebContents,
  sources: LogSource[],
  options: StreamOptions,
): void {
  const normalizedSources = normalizeSources(sources)
  const stream = activeStreams.get(streamId)
  if (!stream || normalizedSources.length === 0) return

  const maxInitialLines = options.maxInitialLines ?? DEFAULT_MAX_INITIAL_LINES
  const initialRecords = queryInitialRecords(normalizedSources, maxInitialLines)
  let lastId = initialRecords.reduce((max, record) => Math.max(max, record.id), 0)

  for (const source of normalizedSources) {
    send(sender, {
      event: 'initial',
      data: {
        source,
        lines: initialRecords.filter((record) => record.source === source).map(toLogLine),
      },
    })
  }

  if (options.tail === false) return

  const poll = async () => {
    const currentStream = activeStreams.get(streamId)
    if (!currentStream || currentStream.cancel || currentStream.pollInFlight) return
    currentStream.pollInFlight = true
    try {
      const newRecords = queryRecordsAfterId(normalizedSources, lastId, DEFAULT_MAX_QUERY_LINES)
      if (newRecords.length === 0) return
      lastId = newRecords[newRecords.length - 1]?.id ?? lastId

      for (const source of normalizedSources) {
        const lines = newRecords.filter((record) => record.source === source).map(toLogLine)
        if (lines.length > 0) {
          send(sender, { event: 'newLines', data: { source, lines } })
        }
      }
    } catch {
      if (currentStream.fallbackStarted) return
      currentStream.fallbackStarted = true
      if (currentStream.interval) {
        clearInterval(currentStream.interval)
        currentStream.interval = undefined
      }
      void runFileStream(streamId, sender, normalizedSources, {
        maxInitialLines,
        tail: true,
      }).catch((fallbackError) => {
        send(sender, {
          event: 'error',
          data: { source: 'app', message: (fallbackError as Error).message },
        })
      })
    } finally {
      currentStream.pollInFlight = false
    }
  }

  stream.interval = setInterval(() => void poll(), DEFAULT_POLL_INTERVAL_MS)
}

async function runFileStream(
  streamId: string,
  sender: WebContents,
  sources: LogSource[],
  options: StreamOptions,
): Promise<void> {
  const maxInitialLines = options.maxInitialLines ?? DEFAULT_MAX_INITIAL_LINES
  const tail = options.tail ?? true
  const intervalMs = DEFAULT_POLL_INTERVAL_MS

  const sourceStates: SourceState[] = []

  for (const source of sources) {
    if (activeStreams.get(streamId)?.cancel) return
    const filePath = sourcePath(source)
    try {
      const lines = await readTailLines(filePath, DEFAULT_MAX_INITIAL_BYTES, maxInitialLines)
      if (activeStreams.get(streamId)?.cancel) return
      const parsed = lines
        .map((line) => parseLine(source, line))
        .filter((line): line is LogLine => line !== null)
      send(sender, { event: 'initial', data: { source, lines: parsed } })
      let statInfo: Awaited<ReturnType<typeof fs.stat>>
      try {
        statInfo = await fs.stat(filePath)
      } catch {
        statInfo = { size: 0 } as Awaited<ReturnType<typeof fs.stat>>
      }
      sourceStates.push({
        source,
        path: filePath,
        position: Number(statInfo.size),
        pending: { value: '' },
      })
    } catch (e) {
      send(sender, {
        event: 'error',
        data: { source, message: (e as Error).message },
      })
      sourceStates.push({ source, path: filePath, position: 0, pending: { value: '' } })
    }
  }

  if (!tail) return

  const stream = activeStreams.get(streamId)
  if (!stream) return

  stream.interval = setInterval(() => {
    if (stream.cancel || stream.pollInFlight) return
    stream.pollInFlight = true
    void (async () => {
      try {
        for (const state of sourceStates) {
          if (stream.cancel) return
          try {
            const { lines, newPos, rotated } = await readNewData(
              state.path,
              state.position,
              state.pending,
            )
            state.position = newPos
            if (lines.length === 0) continue
            const parsed = lines
              .map((line) => parseLine(state.source, line))
              .filter((line): line is LogLine => line !== null)
            send(sender, {
              event: rotated ? 'reset' : 'newLines',
              data: { source: state.source, lines: parsed },
            })
          } catch (e) {
            send(sender, {
              event: 'error',
              data: { source: state.source, message: (e as Error).message },
            })
          }
        }
      } finally {
        stream.pollInFlight = false
      }
    })()
  }, intervalMs)
}

export async function startLogStream(
  sender: WebContents,
  sources: LogSource[],
  options: StreamOptions,
): Promise<string> {
  const streamId = `log-stream-${nextStreamId++}`
  activeStreams.set(streamId, { cancel: false, sender })
  try {
    runDatabaseStream(streamId, sender, sources, options)
  } catch {
    void runFileStream(streamId, sender, sources, options).catch((e) => {
      send(sender, {
        event: 'error',
        data: { source: 'app', message: (e as Error).message },
      })
    })
  }
  return streamId
}

export function stopLogStream(streamId: string): void {
  const stream = activeStreams.get(streamId)
  if (!stream) return
  stream.cancel = true
  if (stream.interval) {
    clearInterval(stream.interval)
  }
  activeStreams.delete(streamId)
}

export async function exportLog(source: LogSource, destPath: string): Promise<void> {
  try {
    const records = queryRecordsForSource(source)
    const content = records.length > 0 ? `${records.map((record) => record.raw).join('\n')}\n` : ''
    await fs.writeFile(destPath, content, 'utf8')
  } catch {
    const src = sourcePath(source)
    await fs.copyFile(src, destPath)
  }
}
